src/eve/pattern_matching.py

```diff
@@ -0,0 +1,139 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import inspect
+from dataclasses import dataclass
+from types import ModuleType
+from typing import Any, Iterator
+
+
+@dataclass(frozen=True)
+class ObjectPattern:
+    cls: type
+    attrs: dict[str, Any]
+
+    def matches(self, other: Any, raise_: bool = False) -> bool:
+        """Return if object pattern matches `other` using :func:`get_differences`.
+
+        If `raise_` is specified raises an exception with all differences found.
+        """
+        diffs = list(get_differences(self, other))
+        if raise_ and len(diffs) != 0:
+            diffs_str = "\n  ".join([f"  {self.cls.__name__}{path}: {msg}" for path, msg in diffs])
+            raise ValueError(f"Object and pattern don't match:\n  {diffs_str}")
+        return len(diffs) == 0
```

Comment:
It doesn't make sense to iterate the generators with all the errors if they haven't been requested by the user. We could just return `False` as soon as the first error is found:

```suggestion
        if not raise_exception:
            return next(get_differences(self, other), None) is None
        else:
            diffs = [*get_differences(self, other)]
            if len(diffs) == 0:
                return True
            else:
                diffs_str = "\n  ".join([f"  {self.cls.__name__}{path}: {msg}" for path, msg in diffs])
                raise ValueError(f"Object and pattern don't match:\n  {diffs_str}")
```

Reply:
True, I've implemented a slightly different, but similar version to what you proposed.